### PR TITLE
Add support for object attributes in Iter call

### PR DESCRIPTION
**Pull request by fpetkovski**
Add support for IterWithAttributes

This commit adds support for an IterWithAttributes on the bucket client.
The method allows iterating through objects and getting multiple attributes
into the callback function, removing the need to do an Iter followed by Attrs.

For now, we only support getting the last updated time as an attribute, but the
implementation allows adding more in the future.

Not all buckets support this method. The client can check whether the bucket has
support by calling the SupportedIterOptions method on the client.

Co-authored-by: Ashwanth Goli <iamashwanth@gmail.com>
Co-authored-by: Filip Petkovski <filip.petkovsky@gmail.com>

Signed-off-by: Filip Petkovski <filip.petkovsky@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

* Add support for `IterWithAttributes`. 

## Verification

* Unit tests.

### CODE REVIEW

File: prefixed_bucket.go [hunk 1]
```diff
@@ -46,11 +46,11 @@ func (p *PrefixedBucket) Close() error {
...
 // object name including the prefix of the inspected directory.
 // Entries are passed to function in sorted order.
-func (p *PrefixedBucket) Iter(ctx context.Context, dir string, f func(string) error, options ...IterOption) error {
+func (p *PrefixedBucket) Iter(ctx context.Context, dir string, f func(name string, attrs ObjectAttributes) error, options ...IterOption) error {
 	pdir := withPrefix(p.prefix, dir)
 
-	return p.bkt.Iter(ctx, pdir, func(s string) error {
-		return f(strings.TrimPrefix(s, p.prefix+DirDelim))
+	return p.bkt.Iter(ctx, pdir, func(s string, _ ObjectAttributes) error {
+		return f(strings.TrimPrefix(s, p.prefix+DirDelim), EmptyObjectAttributes)
```
Comment (brancz) [inline prefixed_bucket.go:53]: why wouldn't it be passed through here?

File: objstore.go [hunk 2]
```diff
@@ -68,10 +70,12 @@ type InstrumentedBucket interface {
...
 type BucketReader interface {
-	// Iter calls f for each entry in the given directory (not recursive.). The argument to f is the full
-	// object name including the prefix of the inspected directory.
+	// Iter calls f for each entry in the given directory (not recursive.). The first argument to f is the full
+	// object name including the prefix of the inspected directory. The second argument are the object attributes
+	// returned by the underlying provider. Attributes can be requested using various IterOption's.
+	//
 	// Entries are passed to function in sorted order.
-	Iter(ctx context.Context, dir string, f func(string) error, options ...IterOption) error
+	Iter(ctx context.Context, dir string, f func(name string, attrs ObjectAttributes) error, options ...IterOption) error
```
Comment (brancz) [inline objstore.go:78]: If we're already breaking this signature, I think we should instead make the whole thing pass _only_ the `ObjectAttributes` struct and add the `Name` field to it.

**Review by brancz — COMMENTED** [review]
I like this change a lot!

File: objstore.go (hunk 2 above)
Comment (fpetkovski) [reply objstore.go:78]: This makes sense, and would allow us to extend attributes in the future.

Comment (brancz) [reply objstore.go:78]: Exactly!

**Comment by fpetkovski** [comment]
Yeah I will look into adding a new APi. I am still not sure how to handle cases where a provider does not support returning attributes.

**Comment by pracucci** [comment]
> Yeah I will look into adding a new APi. I am still not sure how to handle cases where a provider does not support returning attributes.

An idea may be having a function on the objstore client to check if the feature is supported. Then the Iter() contract is that if the feature is supported, attributes will be fetched, otherwise if the feature is not supported it will error out. It's caller responsability to check if the feature is supported by the actual objstore client and take an informed decision about requesting attributes or not.

**Comment by ashwanthgoli** [comment]
@fpetkovski are there plans to revive this pr? :)

Loki project is planning to switch to thanos clients. Having this change will reduce the fanout of listing objects along with metadata which now requires Iter call followed by fetching attributes for all objects

File: objstore.go [hunk 3]
```diff
@@ -101,24 +114,66 @@ type InstrumentedBucketReader interface {
...
+		Apply: func(params *IterParams) {
+			params.Recursive = true
+		},
+	}
+}
+
+// WithUpdatedAt is an option that can be applied to Iter() to
+// include the last modified time in the attributes.
+// NB: Prefixes may not report last modified time.
+// This option is currently supported for the azure, aws, bos, gcs and filesystem providers.
```
Comment (JoaoBraveCoding) [inline objstore.go:147]: nit: just to match the provider name
```suggestion
// This option is currently supported for the azure, s3, bos, gcs and filesystem providers.
```

File: objstore.go [hunk 4]
```diff
@@ -189,6 +244,20 @@ type ObjectAttributes struct {
...
 }
 
+type IterObjectAttributes struct {
+	Name         string
+	lastModified time.Time
+}
+
+func (i *IterObjectAttributes) SetLastModified(t time.Time) {
+	i.lastModified = t
+}
```
Comment (JoaoBraveCoding) [inline objstore.go:254]: Curious question: why did we make assignments to lastModified come through a setter but not name?

**Review by JoaoBraveCoding — APPROVED** [review]
as far as I can see lgtm

File: providers/gcs/gcs.go [hunk 5]
```diff
@@ -181,31 +181,33 @@ func (b *Bucket) Name() string {
...
+
 	// If recursive iteration is enabled we should pass an empty delimiter.
 	delimiter := DirDelim
-	if objstore.ApplyIterOptions(options...).Recursive {
+	if appliedOpts.Recursive {
 		delimiter = ""
 	}
 
-	query := &storage.Query{
+	it := b.bkt.Objects(ctx, &storage.Query{
```
Comment (ashwanthgoli) [inline providers/gcs/gcs.go:207]: @fpetkovski i didn't realise this was only fetching `Name` attribute before.
I think I removed changes from a previous commit of yours, we need to add it back :)

File: objstore.go [hunk 6]
```diff
@@ -546,6 +615,23 @@ func (b *metricBucket) Iter(ctx context.Context, dir string, f func(name string)
 	return err
 }
 
+func (b *metricBucket) IterWithAttributes(ctx context.Context, dir string, f func(IterObjectAttributes) error, options ...IterOption) error {
+	const op = OpIter
+	b.metrics.ops.WithLabelValues(op).Inc()
+
+	err := b.bkt.IterWithAttributes(ctx, dir, f, options...)
```
Comment (ashwanthgoli) [inline objstore.go:622]: updates to `b.metrics.opsDuration` metric is missing

File: objstore.go (hunk 4 above)
Comment (ashwanthgoli) [reply objstore.go:254]: i wanted to make it explicit that `lastModified` might not always be set, made it a private field so the getter contract can return `time.Time, bool`

**Comment by ashwanthgoli** [comment]
thank you! @fpetkovski
i left a couple of comments, but the rest lgtm

File: objstore.go (hunk 6 above)
Comment (ashwanthgoli) [reply objstore.go:626]: LGTM!

**Pull request merged**